src/script/sign.h

```diff
@@ -68,6 +68,392 @@ struct SignatureData {
+            for (auto& key : keypath_pair.second) {
+                s << key;
+            }
+        }
+
+        // Write the number of inputs
+        if (num_ins > 0) {
+            s << PSBT_NUM_IN_VIN;
+            WriteCompactSize(s, num_ins);
+        }
+
+        // Write the unknown things
+        for(auto& entry : unknown) {
+            s << entry.first;
+            s << entry.second;
+        }
+
+        // Separator
+        s << PSBT_SEPARATOR;
+
+        for (unsigned int i = 0; i < tx.vin.size(); i++) {
+            CTxIn in = tx.vin[i];
+            PartiallySignedInput psbt_in = inputs[i];
+            if (in.scriptSig.empty() && in.scriptWitness.IsNull()) {
+                // Write the utxo
+                // If there is a non-witness utxo, then don't add the witness one.
+                if (psbt_in.non_witness_utxo) {
+                    WriteCompactSize(s, sizeof(PSBT_UNSIGNED_TX_NON_WITNESS_UTXO));
+                    s << PSBT_UNSIGNED_TX_NON_WITNESS_UTXO;
+                    WriteCompactSize(s, ::GetSerializeSize(psbt_in.non_witness_utxo, SER_NETWORK, PROTOCOL_VERSION));
+                    s << psbt_in.non_witness_utxo;
+                } else if (!psbt_in.witness_utxo.IsNull()) {
+                    WriteCompactSize(s, sizeof(PSBT_REDEEMSCRIPT_WITNESS_UTXO));
+                    s << PSBT_REDEEMSCRIPT_WITNESS_UTXO;
+                    WriteCompactSize(s, ::GetSerializeSize(psbt_in.witness_utxo, SER_NETWORK, PROTOCOL_VERSION));
+                    s << psbt_in.witness_utxo;
+                }
+
+                // Write any partial signatures
+                for (auto sig_pair : psbt_in.partial_sigs) {
+                    s << SerializeToVector(PSBT_WITNESSSCRIPT_PARTIAL_SIG, Span<const unsigned char>(sig_pair.first.begin(), sig_pair.first.size()));
+                    WriteCompactSize(s, sig_pair.second.size());
+                    s << MakeSpan(sig_pair.second);
+                }
+
+                // Write the sighash type
+                if (psbt_in.sighash_type > 0) {
+                    WriteCompactSize(s, sizeof(PSBT_BIP32_KEYPATH_SIGHASH));
+                    s << PSBT_BIP32_KEYPATH_SIGHASH;
+                    WriteCompactSize(s, sizeof(psbt_in.sighash_type));
+                    s << psbt_in.sighash_type;
+                }
+
+                // Write the index
+                if (use_in_index) {
+                    WriteCompactSize(s, sizeof(PSBT_NUM_IN_VIN));
+                    s << PSBT_NUM_IN_VIN;
+                    WriteCompactSize(s, sizeof(psbt_in.index));
+                    s << psbt_in.index;
+                }
+            }
+
+            // Write unknown things
+            for(auto& entry : psbt_in.unknown) {
+                s << entry.first;
+                s << entry.second;
+            }
+
+            s << PSBT_SEPARATOR;
+        }
+    }
+
+
+    template <typename Stream>
+    inline void Unserialize(Stream& s) {
```

Comment:
In commit "Implement PSBT structures and un/serliazation methods per BIP 174":

General comment on deserialization code: check that keys and values (where relevant) have the expected length. Things like `uint160(key.begin() + 1, key.end())` will assertion fail if the length is wrong.